presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java

```diff
@@ -72,12 +74,12 @@ public static Assignments of(VariableReferenceExpression variable1, RowExpressio
     @JsonCreator
     public Assignments(@JsonProperty("assignments") Map<VariableReferenceExpression, RowExpression> assignments)
     {
-        this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
+        this.assignments = unmodifiableMap(new HashMap<>(requireNonNull(assignments, "assignments is null")));
     }
 
     public List<VariableReferenceExpression> getOutputs()
     {
-        return ImmutableList.copyOf(assignments.keySet());
+        return unmodifiableList(new ArrayList<>(assignments.keySet()));
```

Comment:
I am still a little bit concerned about `getOutput` here can have random order. 
It might be ok during optimization, since we will always lookup by variable name. But during execution, all the variables will be changed to `inputChannel` (i.e. the index in the list returned by getOutputs), so a change in order will cause mismatching. `HashMap`'s `keySet` order can easily change. 

Also, the old code 
```
ImmutableList.copyOf(assignments.keySet());
```
is cheap because `ImmutableList.copyOf` smartly utilized underlying data structure, so no copy actually happened. But, we will do copying now each time. 

Here, probably let's save it to member variable during construction and return its value instead. We needed extra memory but it is not as bad as unexpected behavior and GC.

Reply:
Understood, nice catch!